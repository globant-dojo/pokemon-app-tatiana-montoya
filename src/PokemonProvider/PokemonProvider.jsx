/* External */
import React, { useEffect } from "react";
import { useState } from "react";

const PokemonContext = React.createContext();

const PokemonProvider = (props) => {
  const [pokemonMessages, setPokemonMessages] = useState({});
  useEffect(() => {
    fetch("/languages/es.json")
      .then((res) => res.json())
      .then((data) => {
        setPokemonMessages(data);
      });
  }, []);

  return (
    <PokemonContext.Provider value={{ POKEMON_MESSAGES: pokemonMessages }}>
      {props.children}
    </PokemonContext.Provider>
  );
};

export { PokemonProvider, PokemonContext };
