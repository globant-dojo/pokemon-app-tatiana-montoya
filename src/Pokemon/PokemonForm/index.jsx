/* External */
import React, { useState } from "react";
import Slider from "rc-slider";
import { useDispatch } from "react-redux";
import PropTypes from "prop-types";

/* Components */
import { Input } from "@/components/Input";
import { Button } from "@/components/Button";

/* Actions */
import { savePokemon, editPokemon } from "@/state/pokemon/actions";

/* Styles */
import "rc-slider/assets/index.css";
import "./PokemonForm.css";

/* Other */
import { PokemonContext } from "@/Pokemon/PokemonProvider";

/* Constants */
const INPUT_SIZE = 200;
const SLIDER_STYLES = {
  backgroundColor: "#2771df",
};

export const PokemonForm = ({ setOpenModal, pokemonToEdit, clearPokemon }) => {
  const { POKEMON_MESSAGES } = React.useContext(PokemonContext);

  const [name, setName] = useState(pokemonToEdit.name || "");
  const [image, setImage] = useState(pokemonToEdit.image || "");
  const [attack, setAttack] = useState(pokemonToEdit.attack || 0);
  const [defense, setDefense] = useState(pokemonToEdit.defense || 0);

  const dispatch = useDispatch();

  const onSavePokemon = () => {
    const pokemonInfo = {
      ...pokemonToEdit,
      name,
      image,
      attack,
      defense,
    };

    if (pokemonInfo.id) {
      dispatch(editPokemon(pokemonInfo));
    } else {
      dispatch(savePokemon(pokemonInfo));
    }

    onCloseModal();
  };

  const onCloseModal = () => {
    setOpenModal(false);
    clearPokemon();
  };

  return (
    <form className="pkm-form">
      <p className="pkm-form-p">Pokemon</p>
      <div className="pkm-form-div-info">
        <label className="pkm-form-label">{`${POKEMON_MESSAGES["common.name"]}:`}</label>
        <Input
          size={INPUT_SIZE}
          value={name}
          onChange={(event) => setName(event.target.value)}
        />
        <label className="pkm-form-label">{`${POKEMON_MESSAGES["common.attack"]}:`}</label>
        <div className="pkm-form-div-slider">
          {attack}
          <Slider
            min={0}
            max={100}
            trackStyle={SLIDER_STYLES}
            activeDotStyle={SLIDER_STYLES}
            defaultValue={50}
            value={attack}
            onChange={setAttack}
            className="pkm-form-slider"
          />
          100
        </div>
      </div>
      <div className="pkm-form-div-info">
        <label className="pkm-form-label">{`${POKEMON_MESSAGES["common.image"]}:`}</label>
        <Input
          size={INPUT_SIZE}
          value={image}
          onChange={(event) => setImage(event.target.value)}
        />
        <label className="pkm-form-label">{`${POKEMON_MESSAGES["common.defense"]}:`}</label>
        <div className="pkm-form-div-slider">
          {defense}
          <Slider
            min={0}
            max={100}
            trackStyle={SLIDER_STYLES}
            activeDotStyle={SLIDER_STYLES}
            defaultValue={50}
            value={defense}
            onChange={setDefense}
            className="pkm-form-slider"
          />
          100
        </div>
      </div>
      <div className="pkm-form-div-actions">
        <Button
          aria-label="btn-create"
          iconName="icon-create"
          handleAction={onSavePokemon}
        >
          {POKEMON_MESSAGES["common.save"]}
        </Button>
        <Button
          aria-label="btn-cancel"
          iconName="icon-cancel"
          handleAction={onCloseModal}
        >
          {POKEMON_MESSAGES["common.cancel"]}
        </Button>
      </div>
    </form>
  );
};

PokemonForm.propTypes = {
  setOpenModal: PropTypes.func.isRequired,
  pokemonToEdit: PropTypes.object,
  clearPokemon: PropTypes.func.isRequired,
};
