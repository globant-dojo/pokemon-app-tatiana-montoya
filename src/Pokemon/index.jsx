/* External */
import React, { useState, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import LoadingBar from "react-redux-loading-bar";
import { ToastContainer } from "react-toastify";

/* Components */
import { TopBar } from "@/Pokemon/TopBar";
import { Modal } from "@/components/Modal";
import { PokemonForm } from "@/Pokemon/PokemonForm";
import { PokemonList } from "@/Pokemon/PokemonList";

/* Styles */
import "react-toastify/dist/ReactToastify.css";

/* Actions */
import { getAll } from "@/state/pokemon/actions";

/* Other */
import { PokemonContext } from "@/Pokemon/PokemonProvider";
import { searchPokemonsByKeyword, showMessages } from "@/utils/misc";

export const Pokemon = () => {
  const { POKEMON_MESSAGES } = React.useContext(PokemonContext);

  const dispatch = useDispatch();

  const [openModal, setOpenModal] = useState(false);
  const [pokemonToEdit, setPokemonToEdit] = useState({});
  const [pokemonsSearched, setPokemonsSearched] = useState(null);

  const {
    pokemons: { list: pokemons, status: statusGetPokemons },
    pokemon: {
      save: pokemonSaveReducer,
      edit: pokemonEditReducer,
      delete: pokemonDeleteReducer,
    },
  } = useSelector((store) => store);

  useEffect(() => {
    showMessages(
      dispatch,
      pokemonEditReducer,
      pokemonSaveReducer,
      pokemonDeleteReducer,
      POKEMON_MESSAGES["messages.success"],
      POKEMON_MESSAGES["messages.error"]
    );
  }, [pokemonEditReducer, pokemonSaveReducer, pokemonDeleteReducer]);

  useEffect(() => {
    dispatch(getAll());
  }, []);

  const onEditPokemon = (pokemon) => {
    setPokemonToEdit(pokemon);
    setOpenModal(true);
  };

  const clearPokemonToEdit = () => setPokemonToEdit({});

  const onSearchPokemons = (keyword) => {
    const pokemonsSearched = searchPokemonsByKeyword(pokemons, keyword);
    setPokemonsSearched(pokemonsSearched);
  };

  const pokemonsToShow = pokemonsSearched || pokemons;

  return (
    <>
      <ToastContainer />
      <LoadingBar style={{ backgroundColor: "blue", height: "5px" }} />
      <h3>{POKEMON_MESSAGES["page.title"]}</h3>
      <TopBar setOpenModal={setOpenModal} onSearchPokemons={onSearchPokemons} />
      <PokemonList
        statusGetPokemons={statusGetPokemons}
        pokemons={pokemonsToShow}
        onEditPokemon={onEditPokemon}
        onSearchPokemons={onSearchPokemons}
      />
      {openModal && (
        <Modal>
          <PokemonForm
            setOpenModal={setOpenModal}
            pokemonToEdit={pokemonToEdit}
            clearPokemon={clearPokemonToEdit}
          />
        </Modal>
      )}
    </>
  );
};
