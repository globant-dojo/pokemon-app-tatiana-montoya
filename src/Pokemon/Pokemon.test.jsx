/* External */
import { render } from "@testing-library/react";
import { Provider } from "react-redux";

/* Components */
import { Pokemon } from ".";

/* Other */
import { PokemonContext } from "@/Pokemon/PokemonProvider";
import store from "@/state/configureStore";

describe("Tests on TopBar", () => {
  const renderPokemonApp = () =>
    render(
      <Provider store={store}>
        <PokemonContext.Provider value={{ POKEMON_MESSAGES: {} }}>
          <Pokemon />
        </PokemonContext.Provider>
      </Provider>
    );

  test("it should match with the snapshot", () => {
    const { container } = renderPokemonApp();

    expect(container).toMatchSnapshot();
  });
});
