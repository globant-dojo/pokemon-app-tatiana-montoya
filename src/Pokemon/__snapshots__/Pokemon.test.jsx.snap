// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on TopBar it should match with the snapshot 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    style="direction: ltr;"
  >
    <div
      class=""
      style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 5px; background-color: blue; position: absolute; opacity: 0;"
    />
    <div
      style="display: table; clear: both;"
    />
  </div>
  <h3 />
  <div
    class="pkm-topBar"
  >
    <div
      class="pkm-div-input"
      style="max-width: 300px;"
    >
      <span
        class="icon-search"
      />
      <input
        class="pkm-input"
        placeholder=""
        value=""
      />
    </div>
    <button
      aria-label="btn-new"
      class="pkm-btn"
      type="button"
    >
      <span
        class="icon-add"
      />
    </button>
  </div>
  <h3 />
</div>
`;
